src/stream.rs

```diff
@@ -0,0 +1,104 @@
+use crate::device::Device;
+use crate::utils::mlx_describe;
+
+/// Parameter type for all MLX operations.
+///
+/// Use this to control where operations are evaluated:
+///
+/// If omitted it will use the [default()], which will be [Device::gpu()] unless
+/// set otherwise.
+pub struct StreamOrDevice {
+    stream: Stream,
+}
+
+impl StreamOrDevice {
+    #[allow(dead_code)]
+    fn new(stream: Stream) -> StreamOrDevice {
+        StreamOrDevice { stream }
+    }
+
+    /// The default stream on the default device.
+    ///
+    /// This will be [Device::gpu()] unless [Device::set_default()]
+    /// sets it otherwise.
+    pub fn default() -> StreamOrDevice {
+        StreamOrDevice {
+            stream: Stream::new(),
+        }
+    }
+
+    pub fn new_with_device(device: &Device) -> StreamOrDevice {
+        StreamOrDevice {
+            stream: Stream::default_stream(device),
+        }
+    }
+
+    /// The `[Stream::default_stream()] on the [Device::cpu()]
+    pub fn cpu() -> StreamOrDevice {
+        StreamOrDevice {
+            stream: Stream::default_stream(&Device::cpu()),
+        }
+    }
+
+    /// The `[Stream::default_stream()] on the [Device::gpu()]
+    pub fn gpu() -> StreamOrDevice {
+        StreamOrDevice {
+            stream: Stream::default_stream(&Device::gpu()),
+        }
+    }
+
+    #[allow(dead_code)]
+    fn get_ctx(&self) -> mlx_sys::mlx_stream {
```

Comment:
Maybe also name it `as_ptr`? Or we can rename `Array::as_ptr`. Just thinking we should be consistent in naming these methods that expose the inner c ptr

Reply:
Yeah, agree on consistency. Let's use `as_ptr()` -- usually I'm more in favor of exposing the inner `c_xx` but here its actually the subproperty that has it.

Reply:
Should it be public though?

Reply:
We can get rid of it while we have no uses. I'll delete.